src/core/modules/impl/WarpGatewayContractDefinitionLoader.ts

```diff
@@ -44,16 +45,36 @@ export class WarpGatewayContractDefinitionLoader implements DefinitionLoader {
     this.tagsParser = new TagsParser();
   }
 
+  // Gets ContractDefinition and ContractSource from two caches and returns a combined structure
+  private async getFromCache(cacheKey: string): Promise<ContractDefinition<any> | null> {
+    const contract = await this.cache.get(cacheKey, 'cd');
+    if (!contract) {
+      return null;
+    }
+    const src = await this.cacheSrc.get(contract.cachedValue.srcTxId, 'src');
+    if (!src) {
+        return null
+    }
+    return { ...contract.cachedValue, ...src.cachedValue };
+  }
+
+  // Divides ContractDefinition into entries in two caches to avoid duplicates
+  private async putToCache(contractTxId: string, value: ContractDefinition<any>): Promise<void> {
+    const src = new CacheSrc(value);
+    const contract = new CacheContract(value);
+    await this.cache.put({ contractTxId: contractTxId, sortKey: 'cd' }, contract);
+    await this.cacheSrc.put({ contractTxId: contract.srcTxId, sortKey: 'src' }, src);
+  }
+
   async load<State>(contractTxId: string, evolvedSrcTxId?: string): Promise<ContractDefinition<State>> {
     let cacheKey = contractTxId;
     if (evolvedSrcTxId) {
       cacheKey += `_${evolvedSrcTxId}`;
     }
 
-    const cacheResult = await this.cache.get(cacheKey, 'cd');
-    if (cacheResult) {
+    const result = await this.getFromCache(cacheKey);
```

Comment:
Since the caches has been splitted, I believe we should now also now split the keys and stop using the concatenated `cacheKey`.

I.e.:
I believe this method should now take two parameters:
1. `contractTxId` (for the contractDefinition)
2. `srcTxId` (for the contractSrc) - optional parameter - which should be used if the `evolvedSrcTxId` is set (i.e. the `evolvedSrcTxId` should be passed here)

Then:
1. `getFromCache` should load the src either by the `contractDefinition.srcTxId` or  the `evolvedSrcTxId` - if it has been passed
2. `putToCache` should put the source either under the `contractDefinition.srcTxId` or the `evolvedSrcTxId` - if it has been passed

Not sure if the current implementation would work properly for the evolved source. I'm pretty sure that it would unnecessarily store the same contract definition each the contract source would change.